Add and use a Converter interface and implementations without using BeanUtils

Claudenw opened the pull request:
https://issues.apache.org/jira/browse/CLI-321

Similar to #215 but without using BeanUtils.

Implementation introduces Converter and Verifier interfaces.

Converter does as it says and converts from Str to object.
Verifier verifies that string values are proper before doing the conversion.  Checks are made when the string is added to the values in the Option object.

TypeHandler is reworked to use and manage the mapping of the Converter and Verifier instances to Classes.
Methods are provided to register Converter and Verifier for a class.

Tests added for Converters and Verifiers as well as TypeHandler changes.

Changes to earlier tests:

There was a TypeHandler test that expected the value "3,5" to throw and exception when it was retrieved, that now throws an exception when the value is initially added to the Option during command line parsing.  Test has been updated to account for this.

Some javadocs need to be completed.

garydgregory commented:
Hi @Claudenw
Thank you for the PR.
Run 'mvn' by itself before pushing to catch these build failures.

Claudenw commented:
@garydgregory I think this might be ready for prime time now.

garydgregory commented on src/main/java/org/apache/commons/cli/converters/Verifier.java:
```diff
@@ -0,0 +1,74 @@
+ * @since 1.7
+ */
+@FunctionalInterface
+public interface Verifier {
```
Probably more flexible if this extends `Predicate<String>`.

Claudenw replied:
I think we can get rid of Verifier and just use Predicate<String>

garydgregory replied:
Note that the advantage of a custom interface here is that we are cementing the design to only test `String` input, as opposed `CharSequence` for example.

garydgregory commented on src/main/java/org/apache/commons/cli/converters/Verifier.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * The definition of the functional interface to call when verifying a string
+ * for input Like {@code Predicate<String>} but can throw a RuntimeException.
+ * @since 1.7
```
The next feature version will be `1.7.0` (major.minor.maintenance).

garydgregory commented on src/main/java/org/apache/commons/cli/CommandLine.java:
```diff
@@ -364,29 +365,79 @@ public Object getParsedOptionValue(final char opt) throws ParseException {
+     * @return the value parsed into a particular object.
+     * @throws ParseException if there are problems turning the option value into the desired type
+     * @see PatternOptionBuilder
+     * @since 1.2
```
The next feature version will be `1.7.0` (major.minor.maintenance).

garydgregory commented on src/main/java/org/apache/commons/cli/TypeHandler.java:
```diff
@@ -19,190 +19,306 @@ Licensed to the Apache Software Foundation (ASF) under one or more
      *
-     * @param str the File location
-     * @return The file represented by {@code str}.
+     * @param      str the File location
```
Please revert the formatting changes. It makes the PR noisier, larger and takes longer to review.

garydgregory commented on src/main/java/org/apache/commons/cli/converters/Converter.java:
```diff
@@ -0,0 +1,74 @@
+ * @since 1.7
+ */
+@FunctionalInterface
+public interface Converter<T> {
```
Probably should extend `Function<String, T>`.

Claudenw replied:
It can't since Converter throws a ParseException and Function<String,T> does not allow that.

garydgregory replied:
Right, we don't want to depend on Commons Lang's `FailableException` here.

Next: Throwing `Exception` is an anti-pattern and should be avoided at all costs IMO. Either:

- reconsider and throw runtime exception OR 
- throw a more specific checked exception OR
- parameterize the exception type
-

garydgregory commented on src/test/java/org/apache/commons/cli/OptionTest.java:
```diff
@@ -75,70 +85,69 @@ private static void checkOption(final Option option, final String opt, final Str
-            Integer.class);
-        checkOption(Option.builder().option("a").desc("desc").type(Integer.class).build(), "a", "desc", null, Option.UNINITIALIZED, null, false, false,
-                defaultSeparator, Integer.class);
+        checkOption(Option.builder("a").desc("desc").build(), "a", "desc", null, Option.UNINITIALIZED, null, false,
```
What's changed here? If it's just formatting, please revert, I don't want to take the time to tease apart formatting from feature changes.

garydgregory reviewed (changes requested):
Hello @Claudenw 
Thank you for this PR.
I think the best path is to work toward getting this PR in instead of the BeanUtils version.
Please see my comments, which are few, per my issue formatting.

garydgregory commented:
How about adding support for `Path`?

garydgregory commented:
> @garydgregory Sorry about the formatting.
> 
> What would you want to see for Path?
> 
> I think that Verifier class should become a static class that simply contains static common Predicate instances.
> 
> Where `Verifier` is currently used in the code it will become Predicate .
> 
> I don't think there needs to be a difference. At one time I had the Verifier throwing the ParseException, but that has been fixed, so there is no valid distinction.
> 
> What are your thoughts?
> 
> I also want to move everything in o.a.c.cli.converters into o.a.c.cli

Path: feature parity with File.

garydgregory commented:
> > @garydgregory Sorry about the formatting.
> > What would you want to see for Path?
> > I think that Verifier class should become a static class that simply contains static common Predicate instances.
> > Where `Verifier` is currently used in the code it will become Predicate .
> > I don't think there needs to be a difference. At one time I had the Verifier throwing the ParseException, but that has been fixed, so there is no valid distinction.
> > What are your thoughts?
> > I also want to move everything in o.a.c.cli.converters into o.a.c.cli
> 
> Path: feature parity with File.

Which can be done later.

garydgregory commented:
Hi @Claudenw 

Hm, I think this PR is overly complicated. 

Specifically, I do not think there should be a split between verifying and converting, there should not be a whole `Verifier` infrastructure IMO: The converters do their parsing, and in this PR, the syntaxes between verifying and parsing don't match anyway.

For a simple example, it is legal for an Integer to have a leading `+` character but the regular expression for integers doesn't allow it. 

It's worse if I want to have a Double parameter since the RE for that is large and complex as documented in `Double.valueOf(String)` and it's certainly not what this PR does:
```
     *  final String Digits     = "(\\p{Digit}+)";
     *  final String HexDigits  = "(\\p{XDigit}+)";
     *  // an exponent is 'e' or 'E' followed by an optionally
     *  // signed decimal integer.
     *  final String Exp        = "[eE][+-]?"+Digits;
     *  final String fpRegex    =
     *      ("[\\x00-\\x20]*"+  // Optional leading "whitespace"
     *       "[+-]?(" + // Optional sign character
     *       "NaN|" +           // "NaN" string
     *       "Infinity|" +      // "Infinity" string
     *
     *       // A decimal floating-point string representing a finite positive
     *       // number without a leading sign has at most five basic pieces:
     *       // Digits . Digits ExponentPart FloatTypeSuffix
     *       //
     *       // Since this method allows integer-only strings as input
     *       // in addition to strings of floating-point literals, the
     *       // two sub-patterns below are simplifications of the grammar
     *       // productions from section 3.10.2 of
     *       // The Java Language Specification.
     *
     *       // Digits ._opt Digits_opt ExponentPart_opt FloatTypeSuffix_opt
     *       "((("+Digits+"(\\.)?("+Digits+"?)("+Exp+")?)|"+
     *
     *       // . Digits ExponentPart_opt FloatTypeSuffix_opt
     *       "(\\.("+Digits+")("+Exp+")?)|"+
     *
     *       // Hexadecimal strings
     *       "((" +
     *        // 0[xX] HexDigits ._opt BinaryExponent FloatTypeSuffix_opt
     *        "(0[xX]" + HexDigits + "(\\.)?)|" +
     *
     *        // 0[xX] HexDigits_opt . HexDigits BinaryExponent FloatTypeSuffix_opt
     *        "(0[xX]" + HexDigits + "?(\\.)" + HexDigits + ")" +
     *
     *        ")[pP][+-]?" + Digits + "))" +
     *       "[fFdD]?))" +
     *       "[\\x00-\\x20]*");// Optional trailing "whitespace"
     *
     *  if (Pattern.matches(fpRegex, myString))
     *      Double.valueOf(myString); // Will not throw NumberFormatException
     *  else {
     *      // Perform suitable alternative action
     *  }
```
All of that to say, that you don't have to worry about any of this if you let a converter do its usual parsing work without trying to short-circuit the process with an _additional_ parsing step. IOW, don't parse input twice, we don't need verifiers.

Nit: `@param` Javadoc tags should be listed first.

Claudenw commented:
@garydgregory I have added the Path to the standard converters handled by the TypeHandler and added documentation for the entire span of this change.

I also added a set of getOptionValue() methods that accept a Supplier<String> to create the defaults.

garydgregory reviewed (changes requested):
I'll keep repeating it ;-) We don't need a "verifier". If later, this proves a requirement, we can discuss adding that concept. I still think that "verifying" is doomed to fail with both false positives and false negatives, on top of being redundant, since "verification" is done implicitly when parsing, for example by `Double#parseDouble()`.

Claudenw commented:
In working on Cassandra Stress test tool where we are trying to move to commons-cli there are cases where we want to quickly identify incorrect textual input without having to go through expensive conversion processes.  This also gives an easy way to verify things like enum values and provide better error messages than if you just call `enum.valueOf(str)`.  Providing the option of the verifier means that often the converter is a simple call to an existing method.

garydgregory commented on src/main/java/org/apache/commons/cli/ParseException.java:
```diff
@@ -21,11 +21,33 @@ Licensed to the Apache Software Foundation (ASF) under one or more
+     * Converts any exception except {@code UnsupportedOperationException} to a {@code ParseException}.
+     * if {@code e} is an instance of {@code ParseException} it is returned, otherwise a {@code ParseException} is 
+     * created that wraps it.
+     * <p>
```
Hello @Claudenw 
- Close HTML tags
- Add `@since` tags to new `public` and `protected` elements.

garydgregory commented on src/main/java/org/apache/commons/cli/ParseException.java:
```diff
@@ -21,11 +21,33 @@ Licensed to the Apache Software Foundation (ASF) under one or more
+        if (e instanceof UnsupportedOperationException) {
+            throw (UnsupportedOperationException) e;
+        }
+        
```
Remove extra white-space.

garydgregory commented on src/main/java/org/apache/commons/cli/PatternOptionBuilder.java:
```diff
@@ -102,14 +102,41 @@ public class PatternOptionBuilder {
+
+    /**
+     * Registers custom {@code Converter}s with the {@code TypeHandler}. 
+     */
```
Add `@since` tags to new public and protected elements.

garydgregory commented on src/main/java/org/apache/commons/cli/PatternOptionBuilder.java:
```diff
@@ -102,14 +102,41 @@ public class PatternOptionBuilder {
      *
      * @param ch the specified character
      * @return The class that {@code ch} represents
+     * @deprecated use {@link #getValueType(char)}
```
Add `@Deprecated` annotation to the method.

garydgregory commented on src/main/java/org/apache/commons/cli/PatternOptionBuilder.java:
```diff
@@ -102,14 +102,41 @@ public class PatternOptionBuilder {
+     * @param ch the specified character
+     * @return The class that {@code ch} represents
+     */
+    public static Class<?> getValueType(final char ch) {
```
Add `@since` tags to new public and protected elements.

garydgregory commented on src/main/java/org/apache/commons/cli/TypeHandler.java:
```diff
@@ -19,29 +19,114 @@ Licensed to the Apache Software Foundation (ASF) under one or more
+    }
+
+    /**
+     * Registers a Converter for a Class.  If @code converter} is null registration is cleared for {@code clazz}, and 
```
Only one space after the `.`.

garydgregory commented on src/site/xdoc/usage.xml:
```diff
@@ -384,5 +384,94 @@ catch (ParseException exp) {
     System.out.println("Unexpected exception:" + exp.getMessage());
 }</source>
     </section>
+    <section name="Converting (Parsing) Option Values">
```
Nice! 😄 👏

garydgregory reviewed (changes requested):
@Claudenw 
Thank you for your updates. Please see my comments.

Claudenw commented:
@garydgregory  requested  changes made.  Also found old 'verifier' documentation still in place and removed it.

garydgregory commented on src/main/java/org/apache/commons/cli/TypeHandler.java:
```diff
@@ -19,29 +19,114 @@ Licensed to the Apache Software Foundation (ASF) under one or more
+     */
+    public static void resetConverters() {
+        converterMap.clear();
+        converterMap.put(Object.class, Converter.OBJECT);
```
Hi @Claudenw,
We have a mix of using constants and expressions to define this map. I think it would be better to do it one way or another: Use all constants or use all expressions.
The simplest would be to inline the constants from `Converter` and decrease the new public footprint. 
Any thoughts on this?
